Add PSP port

sharkwouter opened the pull request:
This is still WIP at the moment, but I wanted this to be a bit more visible.

sharkwouter committed:
Initial attempt at port

sharkwouter committed:
Fix window size on PSP

sharkwouter committed:
Create frame for PSP

sharkwouter committed:
Allow enabling sound for PSP

sharkwouter committed:
Render psp frame

sharkwouter committed:
Start SDL mixer with a supported format on PSP

sharkwouter committed:
Merge branch 'master' into psp-port

sharkwouter commented on src/sdl12main.c:
```diff
@@ -340,8 +342,8 @@ int main(int argc, char** argv)
     SDL_CHECK(screen = SDL_SetVideoMode(400, 240, 32, videoflag));
-#elif defined (__PSP__)
-    SDL_CHECK(screen = SDL_SetVideoMode(480 * scale, 272 * scale, 32, videoflag));
+#elif defined (__PSP__) 
```
```suggestion
#elif defined (__PSP__)
```

sharkwouter committed:
Remove uneccesary space

sharkwouter commented:
So where this is at right now is that the frame around the game keeps flickering and sound does not work. Otherwise the game does work. I'm not sure why it's broken. I think the other ports are written with the assumption that if nothing is drawn to a part of the screen, the GPU does not clear those pixels, which the PSP does not do.

FtZPetruska commented:
> So where this is at right now is that the frame around the game keeps flickering and sound does not work. Otherwise the game does work. I'm not sure why it's broken. I think the other ports are written with the assumption that if nothing is drawn to a part of the screen, the GPU does not clear those pixels, which the PSP does not do.

I can't speak for the ngage SDL port, but for the 3DS since we're stuck with software rendering, the frame gets added to the window's surface and implicitly gets copied every time into the framebuffer. You could try adding a copy of the frame in the rendering loop to see if it fixes flickering.

As to audio, I can't test on my end (unless you can run PSP homebrew on my PSTV). Otherwise:
- Ensure `CELESTE_P8_ENABLE_AUDIO` is set to ON (by default it's OFF).
- Ensure your SDL_mixer build does not use `STB` backend for OGG playback (although that doesn't seem to be the case).
- I'm not an expert when it comes to audio playback, but is this workaround necessary and have you tried without it ? I could see how format mismatch would stop playback.
```cpp
#ifdef __PSP__
    if (Mix_OpenAudio(44100, MIX_DEFAULT_FORMAT, 2, 1024) < 0)
#else
    if (Mix_OpenAudio(22050, AUDIO_S16SYS, 1, 1024) < 0)
#endif
```

FtZPetruska committed:
Merge 'upstream/master' into psp-port.

FtZPetruska committed:
PSP: Fix audio playback.

Music playback would not work because the device was opened in stereo.
Audio frequency also needs to be different on PSP if you care about
your ears.

Lastly, libxmp is added to mixer libs as there would be undefined
references using the pspdev/pspdev docker image.

FtZPetruska committed:
PSP: "Fix" frame flickering.

FtZPetruska committed:
PSP: Add a CI build.

FtZPetruska committed:
PSP: Add icon to executable.

FtZPetruska committed:
PSP: Pack files better in CI.

FtZPetruska commented:
@sharkwouter I think I figured out audio playback and the frame issue, the latter is quite hacky tho.

For audio, the snippet I mentioned in my previous message was indeed blocking music playback (not sound effects). All that was needed was to bump down to mono. Since the only difference with the other call was the audio frequency, I refactored the frequency in a `#define`.

For the frame, since the PSP uses double-buffering... well I render it once in both framebuffers. Quite a bit hack, but it works.

Could you try out the latest changes on real hardware ? I only tested it in PPSSPP.
CI now builds and packs the result, [here's the result for 595c8f2](https://github.com/ngagesdk/ccleste/actions/runs/3332904431) for example.

sharkwouter commented:
Oh thanks, I'll do a test tonight then.

I'm kind of surprised setting it to stereo made it not render. SDL should be able to deal with that.

I do feel like there has to be a better solution for the frame rendering hack, but I guess it works for now.

mupfdev commented:
Please let me know if everything works as intended. I will merge it afterwards. Also please add the PSP to the list of supported platforms in the README.md. Compile instructions are optional but appreciated.

sharkwouter committed:
Set frame size correctly

sharkwouter committed:
Update README to include PSP instructions

sharkwouter commented:
This is ready for review now. It works. With sound it's really slow.

FtZPetruska reviewed (approved):
Looks good!

@mupfdev Any idea about the codacy issue ? It complains about formatting after the addition of the PSP in this list:

```md
It has been confirmed to work on the following platforms:

- Linux
- Nintendo 3DS
- Nokia N-Gage
- Windows
- Playstation Portable
```

But my markdown formatter ([prettier](https://prettier.io/)) would not change that part of the file.

mupfdev commented:
Yes, no worries. I had to adjust a setting after moving the repository to the ngagesdk organisation. It should be fine after the next run. Unfortunately I wasn't able to trigger a new one manually.
